aeneas/src/mach/MachLowering.v3

```diff
@@ -1452,28 +1467,51 @@ class MachLowering(mach: MachProgram, compiler: Compiler, config: MachLoweringCo
 		oi.kill();
 		oi.remove();
 	}
+	def genNormRangeGetElem(oi: SsaApplyOp, elem: int) {
+		if (config.ObjectSystem) return void(normId(oi));
+		var inputs = normRefs(oi.inputs);
+		var narr = inputs[0], rangeStart = inputs[1], nindex = inputs[2], rangeType = oi.op.typeArgs[0];
+		var arrayRep = mach.arrayRep(rangeType);
+		var offset = genArrayElemOffset(if(arrayRep.offsets != null, arrayRep.offsets[elem], 0), arrayRep.elemScale, nindex);
+		var pos = ptrAdd(ptrAdd(narr, rangeStart), offset);
+		var machType = mach.machType(arrayRep.getElemElemType(rangeType, elem));
+		var loads = genNormTypedLoads(oi.source, Fact.O_NO_NULL_CHECK, machType, pos, 0);
+		mapN(oi, loads);
+	}
+	def genNormRangeSetElem(oi: SsaApplyOp, elem: int) {
+		if (config.ObjectSystem) return void(normId(oi));
+		var inputs = normRefs(oi.inputs);
+		var narr = inputs[0], rangeStart = inputs[1], nindex = inputs[2], rangeType = oi.op.typeArgs[0];
+		var arrayRep = mach.arrayRep(rangeType);
+		var offset = genArrayElemOffset(if(arrayRep.offsets != null, arrayRep.offsets[elem], 0), arrayRep.elemScale, nindex);
+		var pos = ptrAdd(ptrAdd(narr, rangeStart), offset);
+		var machType = mach.machType(arrayRep.getElemElemType(rangeType, elem));
+		genNormTypedStores(oi, Fact.O_NO_NULL_CHECK, false, machType, pos, 0, inputs, 3);
+		oi.kill();
+		oi.remove();
+	}
 	def genByteArrayGetField(oi: SsaApplyOp, offset: int) {
 		var inputs = normRefs(oi.inputs);
-		var narr = inputs[0], ni_offset = inputs[1];
+		var narr = inputs[0], start = inputs[1];
 		var fieldType = oi.op.typeArgs[0];
 		var nullity = context.compiler.nullity(oi, narr);
 		nullity = genBoundsCheck(oi, nullity).1;
 		var arrayRep = mach.arrayRep(V3.arrayByteType);
 		var arrayOffset = ptrAdd(narr, context.graph.intConst(offset));
-		var arrayElemOffset = genArrayElemOffset(arrayRep.getElemElemOffset(0), 1, ni_offset);
+		var arrayElemOffset = if(oi.op.typeArgs[1] == V3Range.START_TYPE, start, genArrayElemOffset(arrayRep.getElemElemOffset(0), 1, start));
 		var machType = mach.machType(fieldType);
 		var loads = genNormTypedLoadsByteRegion(oi.source, machType, ptrAdd(arrayOffset, arrayElemOffset), 0);
 		mapN(oi, loads);
 	}
 	def genByteArraySetField(oi: SsaApplyOp, offset: int) {
 		var inputs = normRefs(oi.inputs);
-		var narr = inputs[0], ni_offset = inputs[1], nvalue = inputs[2];
+		var narr = inputs[0], start = inputs[1], nvalue = inputs[2];
```

Comment:
Can you keep the `ni_` prefix here for now? I think I want to move all of this code to use `i_` or `ni_` prefixes for all variables that refer to graph nodes.